File: database_integration/enhance_db/src/bin/36_add_polities_cliopatria.rs
/// Add polity_cliopatria column to individuals_regions_cliopatria.
/// Matches individuals to Cliopatria polities by wikipedia_url.
/// Strips parentheses from polity names.
/// Then adds a count column with the number of individuals per polity.
use anyhow::Result;
use indicatif::{ProgressBar, ProgressStyle};
use rusqlite::{params, Connection};
use std::collections::HashMap;
use std::fs;
use std::io::Write;

const DB_PATH: &str = "data/humans_clean.sqlite3";
const CLIO_DB_PATH: &str = "cliopatria_data/processing/data/cliopatria.db";
const TASK_LOG: &str = "task.log";
const BATCH_SIZE: usize = 50_000;

fn log(msg: &str) {
    println!("{}", msg);
    let mut f = fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(TASK_LOG)
        .unwrap();
    writeln!(f, "{}", msg).unwrap();
}

/// Remove surrounding parentheses from polity name: "(British Empire)" -> "British Empire"
fn strip_parens(name: &str) -> String {
    let trimmed = name.trim();
    if trimmed.starts_with('(') && trimmed.ends_with(')') {
        trimmed[1..trimmed.len() - 1].to_string()
    } else {
        trimmed.to_string()
    }
}

fn main() -> Result<()> {
    let _ = fs::remove_file(TASK_LOG);
    log("=== Step 36: Add polities_cliopatria to individuals_regions_cliopatria ===");

    // ========================================================
    // PHASE 1: Build polity lookup from Cliopatria DB
    // ========================================================
    log("[36] Reading Cliopatria polities...");
    let clio_conn = Connection::open(CLIO_DB_PATH)?;
    let mut url_to_polity: HashMap<String, String> = HashMap::new();
    {
        let mut stmt =
            clio_conn.prepare("SELECT name, wikipedia_url FROM polities WHERE wikipedia_url IS NOT NULL")?;
        let rows = stmt.query_map([], |r| {
            Ok((r.get::<_, String>(0)?, r.get::<_, String>(1)?))
        })?;
        for r in rows {
            let (name, url) = r?;
            let clean_name = strip_parens(&name);
            url_to_polity.insert(url, clean_name);
        }
    }
    drop(clio_conn);
    log(&format!(
        "[36] Cliopatria polity lookup: {} entries",
        url_to_polity.len()
    ));

    // ========================================================
    // PHASE 2: Add column to individuals_regions_cliopatria
    // ========================================================
    let conn = Connection::open(DB_PATH)?;
    conn.execute_batch(
        "PRAGMA journal_mode=WAL; PRAGMA synchronous=NORMAL; PRAGMA cache_size=-2000000;",
    )?;

    // Check if column already exists
    let has_column: bool = {
        let mut stmt = conn.prepare("PRAGMA table_info(individuals_regions_cliopatria)")?;
        let cols: Vec<String> = stmt
            .query_map([], |r| r.get::<_, String>(1))?
            .filter_map(|r| r.ok())
            .collect();
        cols.contains(&"polity_cliopatria".to_string())
    };

    if !has_column {
        log("[36] Adding polity_cliopatria column...");
        conn.execute_batch(
            "ALTER TABLE individuals_regions_cliopatria ADD COLUMN polity_cliopatria TEXT;",
        )?;
    } else {
        log("[36] polity_cliopatria column already exists, resetting...");
        conn.execute_batch(
            "UPDATE individuals_regions_cliopatria SET polity_cliopatria = NULL;",
        )?;
    }

    // Check if count column already exists
    let has_count: bool = {
        let mut stmt = conn.prepare("PRAGMA table_info(individuals_regions_cliopatria)")?;
        let cols: Vec<String> = stmt
            .query_map([], |r| r.get::<_, String>(1))?
            .filter_map(|r| r.ok())
            .collect();
        cols.contains(&"count".to_string())
    };

    if !has_count {
        log("[36] Adding count column...");
        conn.execute_batch(
            "ALTER TABLE individuals_regions_cliopatria ADD COLUMN count INTEGER;",
        )?;
    } else {
        log("[36] count column already exists, resetting...");
        conn.execute_batch("UPDATE individuals_regions_cliopatria SET count = NULL;")?;
    }

    // ========================================================
    // PHASE 3: Update polity_cliopatria by matching URL
    // ========================================================
    let total: i64 = conn.query_row(
        "SELECT COUNT(*) FROM individuals_regions_cliopatria",
        [],
        |r| r.get(0),
    )?;
    log(&format!("[36] Total rows to process: {}", total));

    let pb = ProgressBar::new(total as u64);
    pb.set_style(
        ProgressStyle::default_bar()
            .template("{msg} [{bar:40}] {pos}/{len} ({eta})")
            .unwrap(),
    );
    pb.set_message("Matching polities");

    let mut offset: i64 = 0;
    let mut matched = 0u64;
    let mut unmatched = 0u64;

    loop {
        let mut batch: Vec<(String, String)> = Vec::with_capacity(BATCH_SIZE);
        {
            let mut stmt = conn.prepare_cached(
                "SELECT wikidata_id, url FROM individuals_regions_cliopatria
                 ORDER BY rowid
                 LIMIT ?1 OFFSET ?2",
            )?;
            let rows = stmt.query_map(params![BATCH_SIZE as i64, offset], |r| {
                Ok((r.get::<_, String>(0)?, r.get::<_, String>(1)?))
            })?;
            for r in rows {
                batch.push(r?);
            }
        }

        if batch.is_empty() {
            break;
        }

        conn.execute_batch("BEGIN TRANSACTION;")?;
        {
            let mut update = conn.prepare_cached(
                "UPDATE individuals_regions_cliopatria SET polity_cliopatria = ?1 WHERE wikidata_id = ?2",
            )?;

            for (wikidata_id, url) in &batch {
                if let Some(polity) = url_to_polity.get(url.as_str()) {
                    update.execute(params![polity, wikidata_id])?;
                    matched += 1;
                } else {
                    unmatched += 1;
                }
            }
        }
        conn.execute_batch("COMMIT;")?;

        pb.inc(batch.len() as u64);
        offset += batch.len() as i64;

        if offset % 500_000 < BATCH_SIZE as i64 {
            log(&format!(
                "[36] Progress: {}/{} processed, {} matched, {} unmatched",
                offset, total, matched, unmatched
            ));
        }
    }
    pb.finish();

    log(&format!(
        "[36] Polity matching done: {} matched, {} unmatched",
        matched, unmatched
    ));

    // ========================================================
    // PHASE 4: Compute counts per polity and update
    // ========================================================
    log("[36] Computing individuals count per polity...");
    let mut polity_counts: HashMap<String, i64> = HashMap::new();
    {
        let mut stmt = conn.prepare(
            "SELECT polity_cliopatria, COUNT(*) FROM individuals_regions_cliopatria
             WHERE polity_cliopatria IS NOT NULL
             GROUP BY polity_cliopatria",
        )?;
        let rows = stmt.query_map([], |r| {
            Ok((r.get::<_, String>(0)?, r.get::<_, i64>(1)?))
        })?;
        for r in rows {
            let (polity, cnt) = r?;
            polity_counts.insert(polity, cnt);
        }
    }
    log(&format!(
        "[36] Distinct polities with individuals: {}",
        polity_counts.len()
    ));

    log("[36] Updating count column...");
    conn.execute_batch("BEGIN TRANSACTION;")?;
    {
        let mut update = conn.prepare(
            "UPDATE individuals_regions_cliopatria SET count = ?1 WHERE polity_cliopatria = ?2",
        )?;
        for (polity, cnt) in &polity_counts {
            update.execute(params![cnt, polity])?;
        }
    }
    conn.execute_batch("COMMIT;")?;

    // ========================================================
    // PHASE 5: Create index and final stats
    // ========================================================
    log("[36] Creating indexes...");
    conn.execute_batch(
        "CREATE INDEX IF NOT EXISTS idx_irc_polity ON individuals_regions_cliopatria(polity_cliopatria);",
    )?;

    let total_with_polity: i64 = conn.query_row(
        "SELECT COUNT(*) FROM individuals_regions_cliopatria WHERE polity_cliopatria IS NOT NULL",
        [],
        |r| r.get(0),
    )?;
    let total_without: i64 = conn.query_row(
        "SELECT COUNT(*) FROM individuals_regions_cliopatria WHERE polity_cliopatria IS NULL",
        [],
        |r| r.get(0),
    )?;

    log("[36] === Final Statistics ===");
    log(&format!("[36] Total rows: {}", total));
    log(&format!("[36] With polity_cliopatria: {}", total_with_polity));
    log(&format!("[36] Without polity_cliopatria: {}", total_without));

    // Top 20 polities
    let mut top = conn.prepare(
        "SELECT polity_cliopatria, count FROM individuals_regions_cliopatria
         WHERE polity_cliopatria IS NOT NULL
         GROUP BY polity_cliopatria
         ORDER BY count DESC LIMIT 20",
    )?;
    let rows: Vec<(String, i64)> = top
        .query_map([], |r| Ok((r.get(0)?, r.get(1)?)))?
        .filter_map(|r| r.ok())
        .collect();
    log("[36] Top 20 polities by individual count:");
    for (polity, cnt) in &rows {
        log(&format!("[36]   {} -> {}", polity, cnt));
    }

    log("=== Step 36 complete ===");
    Ok(())
}

File: scripts/database_integration_scripts/legacy/36_add_polities_cliopatria.py
"""36 - Add polity_cliopatria + count columns to individuals_regions_cliopatria.

Mirrors `enhance_db/src/bin/36_add_polities_cliopatria.rs`.

  Inputs : individuals_regions_cliopatria (created by 35)
           cliopatria.db / polities (name, wikipedia_url)
  Output : two new columns on individuals_regions_cliopatria:
           polity_cliopatria (matched name, parens stripped) and count.

Usage
-----
    python3 36_add_polities_cliopatria.py            # synthetic
    python3 36_add_polities_cliopatria.py --full     # real DB + cliopatria
"""

from __future__ import annotations

import sqlite3
import tempfile
from pathlib import Path

from tqdm import tqdm

from common import (
    DB_PATH,
    PROJECT_ROOT,
    add_column_if_missing,
    insert_rows,
    log,
    open_db,
    parse_run_mode,
)

CLIO_DB_PATH = PROJECT_ROOT / "cliopatria_data" / "processing" / "data" / "cliopatria.db"
BATCH_SIZE = 50_000


def _strip_parens(name: str) -> str:
    s = name.strip()
    if s.startswith("(") and s.endswith(")"):
        return s[1:-1]
    return s


def run(conn: sqlite3.Connection, clio_db_path: Path | str = CLIO_DB_PATH) -> int:
    log("[DB] 36: Adding polity_cliopatria column...")

    url_to_polity: dict[str, str] = {}
    with sqlite3.connect(str(clio_db_path)) as clio:
        for name, url in clio.execute(
            "SELECT name, wikipedia_url FROM polities WHERE wikipedia_url IS NOT NULL"
        ):
            url_to_polity[url] = _strip_parens(name)
    log(f"[36] url->polity: {len(url_to_polity)}")

    if add_column_if_missing(conn, "individuals_regions_cliopatria", "polity_cliopatria", "TEXT"):
        log("[36] added polity_cliopatria column")
    else:
        conn.execute("UPDATE individuals_regions_cliopatria SET polity_cliopatria = NULL")
    if add_column_if_missing(conn, "individuals_regions_cliopatria", "count", "INTEGER"):
        log("[36] added count column")
    else:
        conn.execute("UPDATE individuals_regions_cliopatria SET count = NULL")
    conn.commit()

    total = conn.execute("SELECT COUNT(*) FROM individuals_regions_cliopatria").fetchone()[0]
    cursor = conn.execute(
        "SELECT wikidata_id, url FROM individuals_regions_cliopatria ORDER BY rowid"
    )
    matched = 0
    buf: list[tuple] = []
    for wid, url in tqdm(cursor, total=total, desc="36_match"):
        polity = url_to_polity.get(url)
        if polity is None:
            continue
        buf.append((polity, wid))
        if len(buf) >= BATCH_SIZE:
            conn.executemany(
                "UPDATE individuals_regions_cliopatria SET polity_cliopatria = ? WHERE wikidata_id = ?",
                buf,
            )
            conn.commit()
            matched += len(buf)
            buf.clear()
    if buf:
        conn.executemany(
            "UPDATE individuals_regions_cliopatria SET polity_cliopatria = ? WHERE wikidata_id = ?",
            buf,
        )
        conn.commit()
        matched += len(buf)
    log(f"[36] matched {matched}")

    counts: dict[str, int] = {}
    for p, c in conn.execute(
        "SELECT polity_cliopatria, COUNT(*) FROM individuals_regions_cliopatria "
        "WHERE polity_cliopatria IS NOT NULL GROUP BY polity_cliopatria"
    ):
        counts[p] = c
    conn.executemany(
        "UPDATE individuals_regions_cliopatria SET count = ? WHERE polity_cliopatria = ?",
        [(c, p) for p, c in counts.items()],
    )
    conn.execute(
        "CREATE INDEX IF NOT EXISTS idx_irc_polity ON individuals_regions_cliopatria(polity_cliopatria)"
    )
    conn.commit()
    return matched


def _sample_main() -> None:
    with tempfile.TemporaryDirectory() as tmp:
        clio = Path(tmp) / "clio.db"
        db = Path(tmp) / "sample.sqlite3"
        with sqlite3.connect(clio) as c:
            c.execute("CREATE TABLE polities (id INTEGER, name TEXT, wikipedia_url TEXT)")
            insert_rows(c, "polities", [
                {"id": 1, "name": "(British Empire)", "wikipedia_url": "https://en.wikipedia.org/wiki/British_Empire"},
                {"id": 2, "name": "France", "wikipedia_url": "https://en.wikipedia.org/wiki/France"},
            ])
        with sqlite3.connect(db) as seed:
            seed.execute(
                "CREATE TABLE individuals_regions_cliopatria (wikidata_id TEXT PRIMARY KEY, "
                "name_en TEXT, url TEXT, origin TEXT)"
            )
            insert_rows(seed, "individuals_regions_cliopatria", [
                {"wikidata_id": "Q1", "name_en": "A", "url": "https://en.wikipedia.org/wiki/France", "origin": "nationality"},
                {"wikidata_id": "Q2", "name_en": "B", "url": "https://en.wikipedia.org/wiki/British_Empire", "origin": "nationality"},
                {"wikidata_id": "Q3", "name_en": "C", "url": "https://en.wikipedia.org/wiki/Nowhere", "origin": "deathplace"},
            ])
        with open_db(db) as conn:
            n = run(conn, clio_db_path=clio)
            rows = conn.execute(
                "SELECT wikidata_id, polity_cliopatria, count FROM individuals_regions_cliopatria"
            ).fetchall()
        log(f"[sample] matched={n}")
        for r in rows:
            log(f"  {r}")


if __name__ == "__main__":
    if parse_run_mode() == "full":
        with open_db() as conn:
            run(conn)
    else:
        _sample_main()
